### PR TITLE
Fix: added key prop to tag and allowed custom key for tags

### LINKED ISSUE #7585: Set key on <img> inside gatsby-image
The `key` should be set on components so that React knows to reset the image when the props change. `gatsby-image` doesn't currently do this which means if you try passing in different props to the same `gatsby-image` component, it'll briefly show the old version of the image before showing the new one.

Probably the simplest way to do this is to set the `src` prop passed into the component as the `key` on the outermost layer.

https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby-image/src/index.js

### PR DESCRIPTION
**Pull request by cweeii**
Closes #7585, closes #7587, and closes #7606

### CODE REVIEW

**Comment by jlengstorf** [comment]
@cweeii We forgot to add the `key` to props validation. 🤦‍♂️

**Comment by cweeii** [comment]
@jlengstorf Ah yes!  I'll add that right now :)

**Comment by cweeii** [comment]
@jlengstorf I changed the key prop name because I think it's actually a special prop: https://reactjs.org/warnings/special-props.html

File: packages/gatsby-image/src/index.js [hunk 1]
```diff
@@ -341,6 +342,7 @@ class Image extends React.Component {
           className={`${className ? className : ``} gatsby-image-wrapper`}
           style={divStyle}
           ref={this.handleRef}
+          key={this.props.tagKey || `fixed-${JSON.stringify(image.srcSet)}`}
```
Comment (DSchau) [inline packages/gatsby-image/src/index.js]: One nit: I might recommend this is added at the destructuring props level, i.e. [here](https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby-image/src/index.js#L194-L205)

Nice way of not duplicating, and it's where the other props are declared too!

**Comment by jlengstorf** [comment]
Hrmm... That would require updating the docs. But I guess this brings up a question: would the key need to be passed down in this case?

For example, if we set:

```js
import Img from 'gatsby-image';

export default props => <Img key="my-image" {...props} />
```

That would _already_ have a key that React would use. So... it wouldn't need to be passed down.

So maybe we should just drop that altogether.

@DSchau @KyleAMathews  I feel like I might be getting lost on details here. Can you check this for unnecessary complexity?

**Comment by DSchau** [comment]
@jlengstorf good point: I think `key` can also be used as a hint, so I don't think @KyleAMathews was thinking that it'd be passed in by the dev. Setting it up ourselves (i.e. removing the `this.props.tagKey ||`), is probably more in line with what we're thinking!

**Comment by cweeii** [comment]
@DSchau @jlengstorf Oh!  That makes sense.  I'll get rid of that prop right now.

**Comment by DSchau** [comment]
@cweeii thanks for being so receptive to feedback :) We'll get there! 🙌

**Comment by DSchau** [comment]
@cweeii @jlengstorf thanks for the contribution 💪 

```
Successfully published:
 - gatsby-image@2.0.12
```

🎉

**Comment by cweeii** [comment]
Awesome, thanks for the help @DSchau @jlengstorf !

**Pull request merged**